modules/dev-tools/deprecation.php

```diff
@@ -199,25 +199,189 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
-				$this->soft_deprecated_notices[ $entity ] = [
-					$version,
-					$replacement,
-				];
+		if ( $this->should_print_deprecated( $version, $base_version ) || $this->should_console_log_deprecated() ) {
+			$backtrace = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS );
+
+			$external_sources = $this->find_third_party_sources( $backtrace );
+
+			$source_message = 'Elementor';
+			$plugin_name = 'unknown';
+			if ( ! ( empty( $external_sources ) ) ) {
+				$calling_source = $external_sources[ array_key_first( $external_sources ) ];
+				$plugin_name = $this->get_plugin_name( $calling_source['file'] );
+				if ( empty( $plugin_name ) ) {
+					$plugin_name = 'none';
+				}
+				$source_message = sprintf( '%s on file %s:%d.', $calling_source['function'], $calling_source['file'], $calling_source['line'] );
+			}
+			if ( $this->should_console_log_deprecated() ) {
+				// Soft deprecated.
+				if ( ! isset( $this->soft_deprecated_notices[ $entity ] ) ) {
+					$this->soft_deprecated_notices[ $entity ] = [
+						$version,
+						$replacement,
+						$source_message,
+					];
+				}
 			}
 
-			if ( defined( 'ELEMENTOR_DEBUG' ) && ELEMENTOR_DEBUG ) {
-				$print_deprecated = true;
+			if ( $this->should_print_deprecated( $version, $base_version ) ) {
+				$print_deprecated = array(
+					'source' => $source_message,
+					'plugin' => $plugin_name,
+				);
 			}
 		}
-
 		return $print_deprecated;
 	}
 
+	private function get_plugin_name( $filename ) {
+		$plugin_path = substr( strstr( $filename, WP_PLUGIN_DIR ), strlen( WP_PLUGIN_DIR ) );
+		if ( empty( $plugin_path ) ) {
+			return '';
+		}
+		$plugin_name = explode( '/', $plugin_path )[1];
+		return $plugin_name;
+	}
+
+	private function is_elementor_file( $stack_element ) {
+		$filename = $stack_element['file'];
+		return ( strpos( $filename, 'elementor/' ) !== false || strpos( $filename, 'elementor-pro/' ) !== false );
+	}
+
+	private function is_plugin( $stack_element ) {
+		$filename = $stack_element['file'];
+		if ( strpos( $filename, WP_PLUGIN_DIR ) !== false ) {
+			return true;
+		}
+		return false;
+
+	}
+
+	private function find_third_party_sources( $stack_trace ) {
+		$sources = array_filter($stack_trace, function ( $elem ) {
+			return ( ! ( $this->is_elementor_file( ( $elem ) ) ) ) && $this->is_plugin( $elem );
+		} );
+		return $sources;
+	}
+
+	/**
+	 * Checks whether deprecation message should be printed.
+	 * If the user is logged in and has admin privileges, the message will be printed.
+	 * If the user is not logged in, the message will be printed only if the user has WP_DEBUG enabled.
+	 * If the user is logged in but does not have admin privileges, the message will be printed only if the user has ELEMENTOR_DEBUG enabled.
+	 *
+	 *
+	 * @param  mixed $version
+	 * @param  mixed $base_version
+	 * @return bool
+	 */
+	private function should_print_deprecated( $version, $base_version ) {
+
+		$elementor_debug = ( defined( 'ELEMENTOR_DEBUG' ) && ELEMENTOR_DEBUG );
```

Comment:
```suggestion
		$is_elementor_debug = ( defined( 'ELEMENTOR_DEBUG' ) && ELEMENTOR_DEBUG );
```

Reply:
sorry, I disagree. it's exactly elementor_debug, the value of the "define", just undefined-safe